File: threeSum.js
var threeSum = function(nums) {
    let triplets = []
    nums = nums.sort((a,b) => a-b)

    for(let i = 0; i < nums.length-2; i++)
    {    
        let j = i+1
        let k = nums.length - 1
        if((i>0&&nums[i]!=nums[i-1])||i==0){
            while(j<k)
            {
                let sum = nums[i]+nums[j]+nums[k]
                if(sum==0){
                    triplets.push([nums[i], nums[j], nums[k]])
                    while(nums[j]==nums[j+1]) j++
                    j++
                    k--
                }
                else if(sum>0)
                    k--
                else
                    j++
            }
        }
    }
    return triplets
};

File: threeSum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums=sorted(nums)
        triplets = []
        for i in range(len(nums)-2):
            j = i+1
            k = len(nums)-1
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            if nums[i] > 0:
                break

            while j<k:
                s = nums[j]+nums[k]+nums[i]
                if s>0:
                    k-=1
                elif s<0:
                    j+=1
                else:
                    triplets.append([nums[i], nums[j], nums[k]])
                    j += 1
                    k -= 1
                    while j < k and nums[j] == nums[j - 1]:
                        j += 1
                    while j < k and nums[k] == nums[k + 1]:
                        k -= 1
                
        return triplets

